### ISSUE TITLE
Mismatch between npmgraph and `npm install`

### ISSUE THREAD

**Issue by JannesMeyer** [broofa/npmgraph#27]
There seems to be a mismatch in the number of packages reported between this project and `npm install` for some packages.

For example, when I install `jest` with npm in an empty folder I get this output:

```
added 525 packages, and audited 525 packages in 9s
```

Of course, this number includes `jest` itself, so I would expect the number of dependencies to be 524.

However, visualising it on [npm.broofa.com](http://npm.broofa.com/) only shows 474 dependencies. I'm trying to figure out where this discrepancy of 50 dependencies comes from.

**Comment by JannesMeyer** [comment]
Looking at `package-lock.json` I get the impression that some packages are installed multiple times by npm because packages depend on incompatible version ranges.

For example the `braces` package gets installed in version 2.3.2 and 3.0.2.

Assuming that this is the reason for the discrepancy, shouldn't these be counted as two dependencies instead of one?

**Comment by broofa** [comment]
Thanks for the report.  It doesn't appear to be an issue with npmgraph ignoring multiple versions of the same module, however.  That case should be accounted for and, in fact, if you go the Graph tab in the inspector and click the "braces (2)" tag in the "Modules" section, you'll see it highlights the two occurrences of `braces` you mentioned.

![image](https://user-images.githubusercontent.com/164050/97478856-c3ca8c80-190e-11eb-85e5-f76522341e78.png)

**Comment by broofa** [comment]
Okay, digging into this a bit further, it looks like npm is installing multiple versions of the same package in different subdirectories under node_modules.  Looking at the `jest-haste-map` module dependencies (smaller, easier to work with than `jest`), npmgraph shows 147 modules, while `npm install` reports 176 modules.  [This diff shows which modules are "extra" in npm install](https://www.diffchecker.com/QQjm9Ze6).

The left side is from the list of modules that npmgraph logs to console when it fetches a dependency graph.  The right side is a cleaned up version of `grep -R --include "package.json" "\"version\":" node_modules` (lists dir and version of each module directory in `node_modules`).

I'm not sure why the `@types/node` version is different between the two ('need to look into that), but that's not relevant here.  Other than that the only difference is that npm has multiple copies of "define-property", "extend-shallow", and a few other modules.  E.g. `define-property@0.2.5` installed in 5 directories ...

```
$ grep -R --include package.json version node_modules | grep "define-property" | grep 0.2.5
node_modules/expand-brackets/node_modules/define-property/package.json:  "version": "0.2.5"
node_modules/static-extend/node_modules/define-property/package.json:  "version": "0.2.5"
node_modules/snapdragon/node_modules/define-property/package.json:  "version": "0.2.5"
node_modules/class-utils/node_modules/define-property/package.json:  "version": "0.2.5"
node_modules/object-copy/node_modules/define-property/package.json:  "version": "0.2.5"
```

tl;dr - The # reported in npmgraph appears to be correct.

**Comment by JannesMeyer** [comment]
Thanks for your answer!

To be clear, I didn't mean to imply that npmgraph somehow generates an incorrect dependency list. I was merely saying that you should consider including these additional versions in the **total number** of dependencies, so that it matches the npm count.

In my opinion, this would make sense, because for all intents and purposes the versions can be treated as different packages.

- The code could be 100% different between each package version (e.g. v2 could be completely different to v3)
- Each version is parsed separately when required/imported in node
- Assuming you use a bundler like webpack, these extra package versions are all included in the bundle

In my opinion, when npm can't dedupe a dependency it is an extra dependency.

**Comment by JannesMeyer** [comment]
Sorry, I think I slightly misunderstood your reply... I didn't fully realise that the diff you posted was showing that npm is double counting/installing the exact same version of a package into multiple directories. I wasn't aware of that behaviour yet.

This indeed seems to mean that the npm count is slightly misleading rather than npmgraph's.

Anyway, thanks for looking into this. And thanks for creating such a great tool!
